src/vs/editor/common/modes.ts

```diff
@@ -392,7 +392,7 @@ export interface ReferenceProvider {
 	/**
 	 * Provide a set of project-wide references for the given position and document.
 	 */
-	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, token: CancellationToken): Location[] | Thenable<Location[]>;
+	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, progress: (value: Location) => void, token: CancellationToken): Location[] | Thenable<Location[]>;
```

Comment:
> JSON patch seems like a wire format implementation detail. I am not convinced that we want that to leak into the vscode codebase.

I am not 100% sure either, but this is all tied to the LSP proposal. So VS Code will have to add a small JSON Patch library, which would then provide something like an `Operation` class that represents an operation of a partial result and have an `apply()` method on it. The callee would just have to call that on its result variable.

> > always call progress with the not-finished result (what the promise will eventually resolve to)
>
> Possible, but that does result in a lot of overhead for large result sets. It is local so not as bad, but still seems less than ideal.

It would be the same array reference, just with more items. So more overhead only if you have to send it over IPC (it probably has to?). Also, if the consumer happens to not use React, he has no way to find out what exactly changed. 

> As long as each location is uniquely identifiable (they are by dir + file + range), then vscode can be smart enough to have complete locations overwrite the associated PartialLocation if it exists.

TypeScript actually has a concept of `Partial`s. It is a mapped type from an interface to an interface with all properties optional.

I think what you are suggesting is having a class that represents something like an _addition_ to an existing location, identified by an identifier. But partial locations are not always uniquely identifiable, especially since they could just be an empty object. Second, if you think about it that is exactly what the `Operation` I described above would do, just without the need for unique identifiers and not tied to Locations (it would work for every method).